internal/client/db/utils.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package db
+
+import (
+	"fmt"
+	"log"
+	"math"
+
+	"github.com/ChainSafe/gossamer/internal/client/db/columns"
+	"github.com/ChainSafe/gossamer/internal/client/db/metakeys"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/database"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime/generic"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// Meta column. The set of keys in the column is shared by full && light storages.
+const columnMeta = columns.Meta
+
+type finalizedState[H, N any] struct {
+	Hash   H
+	Number N
+}
+
+// Database metadata.
+type meta[H, N any] struct {
+	// Hash of the best known block.
+	BestHash H
+	// Number of the best known block.
+	BestNumber N
+	// Hash of the best finalized block.
+	FinalizedHash H
+	// Number of the best finalized block.
+	FinalizedNumber N
+	// Hash of the genesis block.
+	GenesisHash H
+	// Finalized state, if any
+	FinalizedState *finalizedState[H, N]
+	// Block gap, start and end inclusive, if any.
+	BlockGap *[2]N
+}
+
+// A block lookup key: used for canonical lookup from block number to hash
+type numberIndexKey [4]byte
+
+// Convert block number into short lookup key (LE representation) for
+// blocks that are in the canonical chain.
+//
+// In the current database schema, this kind of key is only used for
+// lookups into an index, NOT for storing header data or others.
+func newNumberIndexKey[N runtime.Number](num N) (numberIndexKey, error) {
```

Comment:
Since num gets immediately casted to a uint32 we should just take this as a param. Additionally, we throw an error if num is > `math.MaxUint32` but `runtime.Number` can be a Uint64, so to me its safer/makes more sense to deal directly with uint32 here

Reply:
see https://github.com/ChainSafe/gossamer/pull/4015#discussion_r1648185674.  I plan to actually support this as a follow up, and make the same change to substrate.

Reply:
Gotcha, with these changes that makes sense. And to double check, this change will occur before this is merged into dev right?

Reply:
Yup!